### PR TITLE
Unordered head block

### PR DESCRIPTION
**Pull request by owen-d**
This PR is a no-op which adds some new (currently unused) code for an unordered head block. This will lay the groundwork for removing the ordering constraint and breaks this code into it's own PR.

### CODE REVIEW

**Review by sandeepsukhani — APPROVED** [review]
LGTM

File: pkg/chunkenc/memchunk.go [hunk 1]
```diff
@@ -350,7 +352,7 @@ func NewByteChunk(b []byte, blockSize, targetSize int) (*MemChunk, error) {
 		// Verify checksums.
 		expCRC := binary.BigEndian.Uint32(b[blk.offset+l:])
 		if expCRC != crc32.Checksum(blk.b, castagnoliTable) {
-			level.Error(util_log.Logger).Log("msg", "Checksum does not match for a block in chunk, this block will be skipped", "err", ErrInvalidChecksum)
+			_ = level.Error(util_log.Logger).Log("msg", "Checksum does not match for a block in chunk, this block will be skipped", "err", ErrInvalidChecksum)
```
Comment (cyriltovena) [inline pkg/chunkenc/memchunk.go:355]: You shouldn't need this does the linter caused you trouble ? Might be just locally.

File: pkg/chunkenc/unordered.go [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+				s = &logproto.Series{
+					Labels: parsedLabels.String(),
+				}
+				series[lhash] = s
+			}
+
+			// []byte here doesn't create allocation because Sum64 has go:noescape directive
+			// It specifies that the function does not allow any of the pointers passed as arguments
+			// to escape into the heap or into the values returned from the function.
+			h := xxhash.Sum64([]byte(line))
```
Comment (cyriltovena) [inline pkg/chunkenc/unordered.go:249]: Recently change this one of us will have to be careful.

File: pkg/chunkenc/unordered.go [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+			return nil
+		},
+	)
+
+	if len(series) == 0 {
+		return iter.NoopIterator
+	}
+	seriesRes := make([]logproto.Series, 0, len(series))
+	for _, s := range series {
+		// todo(ctovena) not sure we need this sort.
```
Comment (cyriltovena) [inline pkg/chunkenc/unordered.go:264]: we don't in the end took me long to come back :)

**Review by cyriltovena — APPROVED** [review]
LGTM Curious to see the rest.

Specially around limitation, and cutting head.

File: pkg/chunkenc/memchunk.go (hunk 1 above)
Comment (owen-d) [reply pkg/chunkenc/memchunk.go:355]: Ah, might be a version mismatch - I usually run the linter outside of docker and it's a bit noisy :)

**Comment by owen-d** [comment]
ref https://github.com/grafana/loki/issues/1544

**Pull request merged**